nixos/lib/test-driver/test_driver/driver.py

```diff
@@ -101,6 +106,122 @@ def __exit__(self, *_: Any) -> None:
             for machine in self.machines:
                 machine.release()
 
+    def rebuild(self, cmd: Optional[str] = None, exe: str = sys.argv[0]) -> None:
+        """
+        Only makes sense when running interactively. Rebuilds the test driver by running `cmd`, or a globally defined `driver.rebuild_cmd`. This should be the same command that built this test driver to begin with. Then uses the new driver at path `exe` to reconfigure this one and redeploy changed machines.
+        """
+
+        if cmd is None:
+            cmd = self.rebuild_cmd
+        if cmd is not None:
+            subprocess.run(cmd, shell=True, check=True)
+
+        tmp_dir = get_tmp_dir()
+
+        new_driver_info = subprocess.check_output(
+            [exe, "--internal-print-update-driver-info-and-exit"],
+            text=True,
+        )
+        (
+            start_scripts,
+            vlans_str,
+            testscript,
+            output_directory,
+        ) = new_driver_info.rstrip().split("\n")
+
+        with self.logger.nested("updating machines"):
+            start_cmds = [
+                NixStartScript(start_script) for start_script in start_scripts.split()
+            ]
+            names = [start_cmd.machine_name for start_cmd in start_cmds]
+
+            # delete machines that are no longer part of the test
+            to_del = []
+            for idx, machine in enumerate(self.machines):
+                if machine.name not in names:
+                    if machine.is_up():
+                        self.logger.warning(
+                            f"{machine.name} removed from the test, but it's running so we're not going to shut it down automatically. Call driver.machines[\"{machine.name}\"].stop() and re-run rebuild() to remove."
```

Comment:
If I understand correctly, with N machines running, we still need to stop all of them before executing `rebuild(...)`. This approach doesn’t seem significantly more advantageous than stopping the entire driver and rerunning it with the `--keep-vm-state` parameter. This way, the interactive shell history and the VMs’ disks remain intact for the next run.

Could you please elaborate on your specific use case and the exact benefits you see from this additional complexity? Understanding this will help evaluate if the added complexity is justified.

Reply:
> If I understand correctly, with N machines running, we still need to stop all of them before executing rebuild(...)

No! This is only if you **delete** a machine from the test, communicating that you no longer want it to be there at all.

If you **change** the config of a machine, it will do the equivalent of running `nixos-rebuild switch` on that machine. Namely, it will build the new configuration and run `<new-configuration-store-path>/system/bin/switch-to-configuration test` on the running VM

Reply:
I expect deleting machines from the test in this context will be quite uncommon. This code is just handling an edge case gracefully. Most of the time you will be modifying existing machines, and that's where this PR shines, since it allows you to redeploy the VMs in-place rather than restart them.